Applications/Spire/Source/Charting/ChartView.cpp

```diff
@@ -460,20 +482,80 @@ void ChartView::resizeEvent(QResizeEvent* event) {
 }
 
 void ChartView::showEvent(QShowEvent* event) {
-  if(m_top_left.m_x == ChartValue() && m_top_left.m_y == ChartValue() &&
-      m_bottom_right.m_x == ChartValue() &&
-      m_bottom_right.m_y == ChartValue()) {
+  if(m_region == Region{}) {
     auto current_time = boost::posix_time::second_clock::local_time();
     auto bottom_right = ChartPoint(ChartValue(current_time),
       ChartValue(Nexus::Money(0)));
     auto top_left = ChartPoint(
       ChartValue(current_time - boost::posix_time::hours(1)),
       ChartValue(Nexus::Money(1)));
-    set_region(top_left, bottom_right);
+    set_region({top_left, bottom_right});
   }
   QWidget::showEvent(event);
 }
 
+ChartView::GapInfo ChartView::update_gaps(std::vector<ChartView::Gap>& gaps,
+    std::vector<Candlestick>& candlesticks, ChartValue start) {
+  auto gap_info = GapInfo{0, ChartValue()};
+  for(auto& candlestick : candlesticks) {
+    auto end = candlestick.GetStart();
+    if(start != end) {
+      gaps.push_back({start, end});
+      gap_info.total_gaps_value += end - start;
+      ++gap_info.gap_count;
+    }
+    start = candlestick.GetEnd();
+  }
+  return gap_info;
+}
+
+QtPromise<ChartView::LoadedData> ChartView::load_data(
+    QtPromise<std::vector<Candlestick>>& promise, LoadedData data,
+    ChartModel* model) {
+  return promise.then([=] (auto result) mutable {
+    auto new_candlesticks = std::move(result.Get());
+    if(!new_candlesticks.empty() && !data.m_candlesticks.empty() &&
```

Comment:
I can remove everything else, but I'll at least need to check if data.m_candlesticks is empty, otherwise it may crash if back() is called on it in the find_if() lambda.

Reply:
Okay simplify as much as possible.